src/topology/ready_events.rs

```diff
@@ -0,0 +1,321 @@
+use std::pin::Pin;
+
+use futures::task::{Context, Poll};
+use futures::{Stream, StreamExt};
+
+use crate::event::{EventArray, EventContainer};
+
+/// An adaptor for pulling an appropriately-sized `EventArray` from the stream.
+///
+/// This combinator will attempt to pull ready `Event`s from the stream
+/// and buffer them into a local array. At most `capacity` `Event`s will
+/// get buffered before an `EventArray` is yielded from the returned
+/// stream. If underlying stream returns Poll::Pending, and collected
+/// chunk is not empty, it will be immediately returned.
+pub struct ReadyEvents<S> {
+    inner: S,
+    buffer: Option<EventArray>,
+    capacity: usize,
+}
+
+impl<S> ReadyEvents<S> {
+    fn push(&mut self, incoming: EventArray) -> Option<EventArray> {
+        match self.buffer.take() {
+            None => {
+                let (result, buffer) = split_array(incoming, self.capacity);
+                self.buffer = buffer;
+                result
+            }
+            Some(buffered) => match (buffered, incoming) {
+                (EventArray::Logs(mut buffered), EventArray::Logs(incoming)) => {
+                    buffered.extend(incoming);
+                    let (result, buffer) = split_buffer(buffered, self.capacity, EventArray::Logs);
```

Comment:
TIOLI: My thought here is that it _may_ be more advantageous to return the entire `buffered` array without extending it _if_ extending it would cause it to be split anyways, as we'd be doing the memory copy only to then allocate another array for the split.

To be fair, though, it's not clear to me if the memory copy is worth taking the hit on in order to more evenly pace the event flow and keep batches evenly sized.  I realize this would also theoretically violate the intent of the combinator -- always return up to N ready events -- but the idea of moving, and then maybe still slicing, and allocating again... it definitely makes me cock my head sideways a little bit.

Reply:
That's worth trying out, yes. I'd prefer to leave that for a follow-up PR so we can measure any impact that might have on the soak tests.

Reply:
I can accept that. 👍🏻